services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/flows/FlowValidationNegativeSpec.groovy

```diff
@@ -0,0 +1,159 @@
+import org.openkilda.messaging.model.SwitchId
+import org.openkilda.testing.model.topology.TopologyDefinition.Switch
+import org.openkilda.testing.service.database.Database
+import org.springframework.beans.factory.annotation.Autowired
+
+import spock.lang.Unroll
+
+
+class FlowValidationNegativeSpec extends BaseSpecification {
+
+    @Autowired
+    PathHelper pathHelper
+
+    @Autowired
+    Database database
+
+
+    Map<SwitchId, SwitchFlowEntries> captureSwitchRules(List<SwitchId> switchIds) {
+        def rules = [:]
+        switchIds.each { switchId -> rules[switchId] = northbound.getSwitchRules(switchId) }
+        return rules
+    }
+
+    Map<SwitchId, List<FlowEntry>> findFlowRules(long cookie, List<SwitchId> switches) {
+        Map<SwitchId, List<FlowEntry>> result = [:]
+        captureSwitchRules(switches).findAll { k, v -> result[k] = v.flowEntries.findAll { it.cookie == cookie } }
+        return result
+    }
+
+    def noDirectLinks(Switch src, Switch dst, List<IslInfoData> links) {
+        def connectingLinks = links.findAll {link ->
+            (link.path[0].switchId == src.dpId) && (link.path[-1].switchId == dst.dpId)
+        }
+        return connectingLinks == []
+    }
+
+    def getNonNeighbouringSwitches() {
+        def islInfoData = northbound.getAllLinks()
+        def switches = topologyDefinition.getActiveSwitches()
+        def differentSwitches = [switches, switches].combinations().unique().findAll {src, dst -> src.dpId != dst.dpId}
+
+        return differentSwitches.find {src, dst -> noDirectLinks(src, dst, islInfoData)}
+    }
+
+    def getNeighbouringSwitches() {
+        def switches = topologyDefinition.getActiveSwitches()
+        def islInfoData = northbound.getAllLinks()
+        return [switches, switches].combinations().unique().find { src, dst -> !noDirectLinks(src, dst, islInfoData) }
+    }
+
+    def getSingleSwitch() {
+        def switches = topologyDefinition.getActiveSwitches()
+        return [switches.first(), switches.first()]
+    }
+
+    @Unroll
+    def "Flow validation should fail if #rule switch rule is deleted"() {
+        given: "Flow with a transit switch exists"
+
+        def (Switch srcSwitch, Switch dstSwitch) = getNonNeighbouringSwitches()
+        assert srcSwitch && dstSwitch
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        flow = northbound.addFlow(flow)
+        assert flow?.id
+        assert Wrappers.wait(20) { "up".equalsIgnoreCase(northbound.getFlow(flow.id).status) }
+        def forward = database.getFlow(flow.id)?.left
+        def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id).collect {it -> it.dpId}
+
+        def newRules = findFlowRules(forward.cookie, involvedSwitches)
+        and: "Rules are installed on all switches"
+        assert newRules.size() > 2
+
+        when: "${rule} switch rule is deleted"
+        def cookie = newRules[newRules.keySet()[item]][0].cookie.toString()
```

Comment:
Omg it literally took me about 40 minutes to understand that this `cookie` is the same as `forward.cookie`. I don't understand why are you trying to obfuscate the simple test.
```
given: "Flow with transit switch"
//everything fine here until assert that flow is up at line 74

and: "Defined forward cookie to be deleted"
def fwCookie = database.getFlow(flow.id).left.cookie

and: "The cookie is present on all switches of the flow"
//by the way I don't think that this verification should be present in this test. Probably should be present 
//in flow crud, but not here. Still adding it to match your original steps
def involvedSwitches = pathHelper.getInvolvedSwitches(flow.id)
assert involvedSwitches.every{ sw ->
   northbound.getSwitchRules(sw.dpId).flowEntries.find { it.cookie == fwCookie }
}

when: "Delete #rules cookie"
northbound.deleteSwitchRules(involvedSwitches[item], fwCookie)

then: "Flow validation reports missing cookie"
def validation = northbound.validateFlow(flow.id)
//verify that only forward flow is broken
//verify that only fwCookie is reported as missing 
??profit
```
This is it. No need for additional mind blowing `Map<SwitchId, List<FlowEntry>> findFlowRules` methods and obfuscated test code

Reply:
If you want we can discuss how this implementation is different from yours and weigh the pros and cons